### PR TITLE
Adding copy/paste provider API proposal

### PR DESCRIPTION
**Pull request by mjbvz**
For #30066

This adds a new `documentCopyPaste` api proposal that lets extensions hook into copy and paste

This can be used to do things such as:

- Create link when pasting an image
- Bring along imports when copy and pasting code

### CODE REVIEW

File: extensions/markdown-language-features/src/extension.ts [hunk 1]
```diff
@@ -53,6 +53,45 @@ export function activate(context: vscode.ExtensionContext) {
 		logger.updateConfiguration();
 		previewManager.updateConfiguration();
 	}));
+
+	// Example copy paste provider that includes the number of times
+	// you've copied something in the pasted text.
+
+	let copyCount = 0;
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/extension.ts:60]: For testing purposes and should not be merged

File: src/vscode-dts/vscode.proposed.documentPaste.d.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		 * Optional method invoked after the user copies text in a file.
+		 *
+		 * During `provideCopyData`, an extension can compute metadata that is attached to
+		 * the clipboard and is passed back to the provider in `providePasteDocument`.
+		 *
+		 * @param document Document where the copy took place.
+		 * @param selection Selection being copied in the `document`.
+		 * @param token
+		 */
+		prepareDocumentPaste?(document: TextDocument, selection: Selection, dataTransfer: DataTransfer, token: CancellationToken): void | Thenable<void>;
```
Comment (mjbvz) [inline src/vscode-dts/vscode.proposed.documentPaste.d.ts:25]: Selection -> Range

File: src/vscode-dts/vscode.proposed.documentPaste.d.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		 * @param dataTransfer
+		 * @param token
+		 *
+		 * @return Optional workspace edit that applies the paste. Return undefined to use standard pasting.
+		 */
+		provideDocumentPasteEdits(document: TextDocument, selection: Selection, dataTransfer: DataTransfer, token: CancellationToken): ProviderResult<WorkspaceEdit>;
+	}
+
+	namespace languages {
+		export function registerDocumentPasteEditProvider(selector: DocumentSelector, provider: DocumentPasteEditProvider): Disposable;
```
Comment (mjbvz) [inline src/vscode-dts/vscode.proposed.documentPaste.d.ts:43]: TODO: add display name to show in UI?

**Review by mjbvz — COMMENTED** [review]
Notes from discussions: 

- Right now we always apply the first edit from a paste provider. In the future we may instead want to show UI on paste that lets users select  which edit to use


- Instead of returning a `WorkspaceEdit` directly, we may want to return a new type such as `PasteAction` that has a `WorkspaceEdit`. This would let extensions provide names if we show the action in the UI, and also potentially let us split the paste into two phases: provide and resolve

- The current implementation does not apply data transfer contributed by extensions to the global clipboard

     This means that you cannot copy and paste data that comes from a provider in VS Code into out into another application. This is likely something we can fix, but means we would need permissions to access the clipboard api

File: src/vs/editor/browser/widget/codeEditorWidget.ts [hunk 4]
```diff
@@ -2033,6 +2033,7 @@ export class EditorModeContext extends Disposable {
 		this._register(_languageFeaturesService.completionProvider.onDidChange(update));
 		this._register(_languageFeaturesService.codeActionProvider.onDidChange(update));
 		this._register(_languageFeaturesService.codeLensProvider.onDidChange(update));
+		this._register(_languageFeaturesService.documentPasteEditProvider.onDidChange(update));
```
Comment (jrieken) [inline src/vs/editor/browser/widget/codeEditorWidget.ts:2036]: Is this needed? I see no paste-related context key that's being updated?

File: src/vs/workbench/api/common/extHost.api.impl.ts [hunk 5]
```diff
@@ -459,6 +459,9 @@ export function createApiFactoryAndRegisterActors(accessor: ServicesAccessor): I
 			registerCodeActionsProvider(selector: vscode.DocumentSelector, provider: vscode.CodeActionProvider, metadata?: vscode.CodeActionProviderMetadata): vscode.Disposable {
 				return extHostLanguageFeatures.registerCodeActionProvider(extension, checkSelector(selector), provider, metadata);
 			},
+			registerDocumentPasteEditProvider(selector: vscode.DocumentSelector, provider: vscode.DocumentPasteEditProvider): vscode.Disposable {
+				return extHostLanguageFeatures.registerDocumentPasteEditProvider(extension, checkSelector(selector), provider);
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHost.api.impl.ts:463]: proposal must be enforced

Comment (jrieken) [reply src/vs/workbench/api/common/extHost.api.impl.ts:463]: and then added to markdown's `package.json`

File: src/vs/editor/contrib/copyPaste/browser/copyPasteController.ts [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+						asString: async () => value,
+						asFile: () => undefined,
+					});
+				}
+			}
+
+			dataTransfer.delete(vscodeClipboardMime);
+
+			for (const provider of providers) {
+				const edit = await provider.provideDocumentPasteEdits(model, selection, dataTransfer, CancellationToken.None);
```
Comment (jrieken) [inline src/vs/editor/contrib/copyPaste/browser/copyPasteController.ts:166]: fyi - there is `EditorStateCancellationTokenSource` which helps you check the document version and automatically cancel requests

**Review by jrieken — CHANGES_REQUESTED** [review]
Looking good, but API proposals need to be enforced 👮🏼‍♂️

**Review by jrieken — APPROVED** [review]

**Pull request merged**